src/pg/pg-store.ts

```diff
@@ -0,0 +1,387 @@
+   * @param limit number of jobs to retrieve
+   * @returns `DbJob[]`
+   */
+  async getPendingJobBatch(args: { limit: number }): Promise<DbJob[]> {
+    return this.sql<DbJob[]>`
+      SELECT * FROM jobs
+      WHERE status = 'pending'
+      ORDER BY updated_at ASC
+      LIMIT ${args.limit}
+    `;
+  }
+
+  /**
+   * Enqueues the tokens specified by a SIP-019 notification for metadata refresh. Depending on the
+   * token type and notification parameters, this will refresh specific tokens or complete
+   * contracts. See SIP-019 for more info.
+   * @param notification SIP-019 notification
+   */
+  async enqueueTokenMetadataUpdateNotification(args: {
+    notification: TokenMetadataUpdateNotification
+  }): Promise<void> {
+    await this.sql.begin(async sql => {
+      // First, make sure we have the specified contract.
+      const contractResult = await sql<{ id: number }[]>`
+        SELECT id FROM smart_contracts WHERE principal = ${args.notification.contract_id}
+      `;
+      if (contractResult.count === 0) {
+        throw new Error(`Contract not found with principal: ${args.notification.contract_id}`);
+      }
+      const contractId = contractResult[0].id;
+
+      if (args.notification.token_class === 'nft') {
+        if (!args.notification.token_ids) {
+          // If this is an NFT update and no token ids were specified, simply re-queue the complete
+          // contract to refresh all tokens.
+          await sql`
+            UPDATE jobs
+            SET status = 'pending', updated_at = NOW()
+            WHERE smart_contract_id = ${contractId}
+          `;
+        } else {
+          // Enqueue each specified token id otherwise.
+          const insertValues: DbTokenInsert[] = args.notification.token_ids.map(i => ({
+            smart_contract_id: contractId,
+            token_number: i,
+            type: DbTokenType.nft,
+          }));
+          await this.getInsertAndEnqueueTokensCursorInternal(insertValues, sql);
+        }
+      } else if (args.notification.token_class === 'ft') {
+        // Enqueue the only token for FTs.
+        await this.getInsertAndEnqueueTokensCursorInternal([{
+          smart_contract_id: contractId,
+          token_number: 1,
+          type: DbTokenType.ft
+        }], sql);
+      }
+    });
+  }
+
+  /**
+   * Returns a token ETag based on its last updated date.
+   * @param contractPrincipal smart contract principal
+   * @param tokenNumber token number
+   * @returns ETag
+   */
+  async getTokenEtag(args: {
+    contractPrincipal: string;
+    tokenNumber: number;
+  }): Promise<string | undefined> {
+    const result = await this.sql<{ etag: string }[]>`
+      SELECT date_part('epoch', t.updated_at)::text AS etag
+      FROM tokens AS t
+      INNER JOIN smart_contracts AS s ON s.id = t.smart_contract_id
+      WHERE s.principal = ${args.contractPrincipal}
+      AND t.token_number = ${args.tokenNumber}
```

Comment:
I'm curious how fast this cache-revalidation query takes in comparison to the regular query for token data. I.e., the cache re-validation needs to be very close to "free" (like the chaintip query in the API). Do you know if this fast enough with a full db (mostly looking at the JOIN)?

Reply:
Good question. It's really quick given that it only pivots on indexed columns (see EXPLAIN below), and in order to get the full metadata result you'd need to make this JOIN + other 3-4 joins (metadata, metadata_attributes, metadata_properties), parsing, and schema validation.

With a full DB, the largest full response I've seen takes like 28ms vs 1.5ms cached.

Here's an EXPLAIN of this query:
```
Nested Loop  (cost=0.70..13.64 rows=1 width=32) (actual time=0.021..0.021 rows=0 loops=1)
  ->  Index Scan using smart_contracts_principal_index on smart_contracts s  (cost=0.28..8.29 rows=1 width=4) (actual time=0.020..0.021 rows=0 loops=1)
        Index Cond: (principal = 'test'::text)
  ->  Index Scan using tokens_smart_contract_id_token_number_unique on tokens t  (cost=0.42..10.22 rows=2 width=12) (never executed)
        Index Cond: ((smart_contract_id = s.id) AND (token_number = 1))
Planning Time: 0.149 ms
Execution Time: 0.035 ms
```

Reply:
Awesome, great profiling 👍